File: android-app/app/src/main/java/com/redcode/im/androidapp/data/contacts/FriendAPIEndpoint.kt
package com.redcode.im.androidapp.data.contacts

import com.redcode.im.androidapp.network.APIEndpoint
import com.redcode.im.androidapp.network.HTTPMethod
import java.net.URLEncoder

object FriendAPIEndpoint {
    val friends = APIEndpoint(HTTPMethod.GET, "/friends")
    val createFriendRequest = APIEndpoint(HTTPMethod.POST, "/friends/requests")

    fun searchUsers(keyword: String, limit: Int = 20): APIEndpoint =
        withQuery(
            "/users/search",
            mapOf(
                "keyword" to keyword.trim(),
                "limit" to limit.coerceIn(1, 50).toString(),
            ),
        )

    fun friendRequests(direction: String? = null, status: String? = null): APIEndpoint =
        withQuery(
            "/friends/requests",
            mapOf("direction" to direction, "status" to status),
        )

    fun respondFriendRequest(requestId: String): APIEndpoint =
        APIEndpoint(HTTPMethod.POST, "/friends/requests/$requestId/respond")

    fun ensurePrivateChat(friendUserId: String): APIEndpoint =
        APIEndpoint(HTTPMethod.POST, "/friends/$friendUserId/chat")

    private fun withQuery(path: String, values: Map<String, String?>): APIEndpoint {
        val query =
            values.entries
                .filter { (_, value) -> !value.isNullOrBlank() }
                .joinToString("&") { (key, value) ->
                    "${encode(key)}=${encode(value.orEmpty())}"
                }
        return APIEndpoint(HTTPMethod.GET, if (query.isBlank()) path else "$path?$query")
    }

    private fun encode(value: String): String =
        URLEncoder.encode(value, Charsets.UTF_8.name())
}

File: ios-app/Sources/RedCodeNetworking/FriendAPIEndpoint.swift
import Foundation

public enum FriendAPIEndpoint: Sendable {
    public static func searchUsers(keyword: String, limit: Int = 20) -> APIEndpoint {
        APIEndpoint(
            method: .get,
            path: "/users/search",
            queryItems: [
                URLQueryItem(name: "keyword", value: keyword),
                URLQueryItem(name: "limit", value: String(limit)),
            ]
        )
    }

    public static let friends = APIEndpoint(method: .get, path: "/friends")
    public static let friendRequests = APIEndpoint(method: .get, path: "/friends/requests")
    public static let createFriendRequest = APIEndpoint(method: .post, path: "/friends/requests")

    public static func friendRequests(direction: String? = nil, status: String? = nil) -> APIEndpoint {
        var queryItems: [URLQueryItem] = []
        if let direction, !direction.isEmpty {
            queryItems.append(URLQueryItem(name: "direction", value: direction))
        }
        if let status, !status.isEmpty {
            queryItems.append(URLQueryItem(name: "status", value: status))
        }
        return APIEndpoint(method: .get, path: "/friends/requests", queryItems: queryItems)
    }

    public static func respondFriendRequest(requestID: String) -> APIEndpoint {
        APIEndpoint(method: .post, path: "/friends/requests/\(requestID)/respond")
    }

    public static func ensurePrivateChat(friendUserID: String) -> APIEndpoint {
        APIEndpoint(method: .post, path: "/friends/\(friendUserID)/chat")
    }

    public static func deleteFriend(friendUserID: String) -> APIEndpoint {
        APIEndpoint(method: .delete, path: "/friends/\(friendUserID)")
    }
}
